blockchain_storage/src/lib.rs

```diff
@@ -26,23 +27,41 @@ impl From<storage::Error> for Error {
 /// Possibly failing result of blockchain storage query
 pub type Result<T> = core::result::Result<T, Error>;
 
-/// Operations on persistent blockchain data
-pub trait BlockchainStorage {
+/// Queries on persistent blockchain data
+pub trait BlockchainStorageRead {
```

Comment:
Is this separation between read and write necessary? This is really unfortunate... Please explain why this is necessary. (I know there's an alias BlockchainStorage that seems to have replaced the old BlockchainStorage and seems to have taken over the same functionality of read/write, but still, is this really necessary?)

Reply:
These methods are on the `TransactionRo` and `TransactionRw` types which implement different subsets of these operations. Maybe we could differentiate between mutable and immutable operations using the `mut` anootation, but I don't think it would work on transaction directly. We would have to introduce an intermediate type and have the transaction types `Deref` and `DerefMut` into that type. I may explore this further but there seems to be a tradeoff: either there are two separate traits or there is an extra itermediate type.

Reply:
Isn't this issue primarily caused by the fact that transactions are either Ro or Rw? I can imagine that write operations can't be attached to Ro because the object expects everything to be immutable.

Reply:
The issue is (kind of) the fact that the `transaction_{ro,rw}` returns a fresh `Transaction{Ro,Rw}` that the caller takes ownership of. Therefore, the caller can make the transaction object mutable or immutable at will. The best we can do as far as I can see is to introduce an extra intermadiate type (I call it `Handle` until a better name shows up). This new type would implement the storage operations via a single trait that combines `BlockchainStorage{Read,Write}`. The transaction types would give access to it by implementing:
```rust
impl Deref for TranctionRo { target = Handle; ... }
impl Deref for TranctionRw { target = Handle; ... }
impl DerefMut for TranctionRw { target = Handle; ... }
```

I have not tried to implement this yet so may need some modifications to work. Also I am not convinced it is a big improvement over having the trait split in two parts. Any ideas / comments?

Reply:
I experimented a bit with the extra type + `Deref` approach and it seems to me it's not really worth it. We can get rid of the `Read/Write` distinction but the extra complexity that crops up elsewhere basically negates the benefits in my view. We can revisit in the future if this causes issues.